main.go

```diff
@@ -1,65 +1,83 @@
+	Wind_MPH float64 `json:"wind_mph"`
 }
 
-var apiKey = os.Getenv("bc94089fef9d46968ff8d724889f9fff")
+func getWeatherByCity(city string) (*WeatherData, error) {
+	url := "https://weatherapi-com.p.rapidapi.com/current.json?q=" + city
 
-func main() {
-	http.HandleFunc("/weather/", func(w http.ResponseWriter, r *http.Request) {
-		city := r.URL.Path[len("/weather/"):]
+	req, _ := http.NewRequest("GET", url, nil)
 
-		url := "https://weatherapi-com.p.rapidapi.com/current.json?q=" + city
+	req.Header.Add("X-RapidAPI-Key", "356fcc1b96msh6b0261faa77fcd2p101d17jsnbbb68ac024b7")
+	req.Header.Add("X-RapidAPI-Host", "weatherapi-com.p.rapidapi.com")
 
-		req, _ := http.NewRequest("GET", url, nil)
+	res, err := http.DefaultClient.Do(req)
 
-		req.Header.Add("X-RapidAPI-Key", "356fcc1b96msh6b0261faa77fcd2p101d17jsnbbb68ac024b7")
-		req.Header.Add("X-RapidAPI-Host", "weatherapi-com.p.rapidapi.com")
+	if err != nil {
+		return nil, errors.New("Error getting weather data")
+	}
 
-		res, err := http.DefaultClient.Do(req)
+	defer res.Body.Close()
+	body, _ := ioutil.ReadAll(res.Body)
 
-		if err != nil {
-			fmt.Println("HTTP call failed:", err)
-			return
-		}
+	var data map[string]interface{}
+    err2 := json.Unmarshal([]byte(string(body)), &data)
 
-		defer res.Body.Close()
-		body, _ := ioutil.ReadAll(res.Body)
+    if err2 != nil {
+        return nil, errors.New("Error getting weather data")
+    }
 
-		weatherData := WeatherData{
-			City: city,
-			Info: string(body),
-		}
-		jsonData, _ := json.Marshal(weatherData)
-		ioutil.WriteFile("weather.json", jsonData, 0600)
+	var weatherData WeatherData
 
-		data, err := ioutil.ReadFile("weather.json")
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
+	if location, ok := data["location"].(map[string]interface{}); ok {
+		if current, ok := data["current"].(map[string]interface{}); ok {
+			weatherData := WeatherData{
+				City: location["name"].(string),
+				Country: location["country"].(string),
+				Temp_C: current["temp_c"].(float64),
+				Temp_F: current["temp_f"].(float64),
+				Wind_KPH: current["wind_kph"].(float64),
+				Wind_MPH: current["wind_mph"].(float64),
+			}
+			return &weatherData, nil
+		} else {
+			return nil, errors.New("Error getting weather data")
 		}
+	} else {
+		return nil, errors.New("Error getting weather data")
+	}
 
-		fmt.Fprintf(w, string(data))
-	})
+	return &weatherData, nil
+}
 
-	server := &http.Server{
-		Addr:              ":8080",
-		ReadHeaderTimeout: 3 * time.Second,
-	}
+func getWeather(context *gin.Context) {
+	city := context.Param("city")
+	weatherData, err := getWeatherByCity(city)
 
-	err := server.ListenAndServe()
 	if err != nil {
-		panic(err)
+		context.IndentedJSON(http.StatusNotFound, gin.H{"error": "Was not able to get weather data"})
+		return
 	}
+
+	context.IndentedJSON(http.StatusOK, weatherData)
 }
+
+func main() {
+	r := gin.Default()
+	r.GET("/:city", getWeather)
+	r.Run("localhost:8080")

```

Comment:
When you shut down the app, how do you make sure requests are finished before the app exits?
E.g. you are running this app in production. You are deploying a new version and are using an orchestration system to ease upgrades (e.g. kubernetes). Is this safe enough to not cut user requests off when you are upgrading your app the to new version, or do you need to do some more magic?